library/certs.c

```diff
@@ -180,26 +180,27 @@ const size_t mbedtls_test_ca_crt_rsa_len = sizeof( mbedtls_test_ca_crt_rsa );
 
 #if defined(MBEDTLS_SHA256_C)
 /* tests/data_files/server2-sha256.crt */
-#define TEST_SRV_CRT_RSA_SHA256                                          \
-"-----BEGIN CERTIFICATE-----\r\n"                                        \
-"MIIDNzCCAh+gAwIBAgIBAjANBgkqhkiG9w0BAQsFADA7MQswCQYDVQQGEwJOTDER\r\n"   \
-"MA8GA1UECgwIUG9sYXJTU0wxGTAXBgNVBAMMEFBvbGFyU1NMIFRlc3QgQ0EwHhcN\r\n"   \
-"MTkwMjEwMTQ0NDA2WhcNMjkwMjEwMTQ0NDA2WjA0MQswCQYDVQQGEwJOTDERMA8G\r\n"   \
-"A1UECgwIUG9sYXJTU0wxEjAQBgNVBAMMCWxvY2FsaG9zdDCCASIwDQYJKoZIhvcN\r\n"   \
-"AQEBBQADggEPADCCAQoCggEBAMFNo93nzR3RBNdJcriZrA545Do8Ss86ExbQWuTN\r\n"   \
-"owCIp+4ea5anUrSQ7y1yej4kmvy2NKwk9XfgJmSMnLAofaHa6ozmyRyWvP7BBFKz\r\n"   \
-"NtSj+uGxdtiQwWG0ZlI2oiZTqqt0Xgd9GYLbKtgfoNkNHC1JZvdbJXNG6AuKT2kM\r\n"   \
-"tQCQ4dqCEGZ9rlQri2V5kaHiYcPNQEkI7mgM8YuG0ka/0LiqEQMef1aoGh5EGA8P\r\n"   \
-"hYvai0Re4hjGYi/HZo36Xdh98yeJKQHFkA4/J/EwyEoO79bex8cna8cFPXrEAjya\r\n"   \
-"HT4P6DSYW8tzS1KW2BGiLICIaTla0w+w3lkvEcf36hIBMJcCAwEAAaNNMEswCQYD\r\n"   \
-"VR0TBAIwADAdBgNVHQ4EFgQUpQXoZLjc32APUBJNYKhkr02LQ5MwHwYDVR0jBBgw\r\n"   \
-"FoAUtFrkpbPe0lL2udWmlQ/rPrzH/f8wDQYJKoZIhvcNAQELBQADggEBAC465FJh\r\n"   \
-"Pqel7zJngHIHJrqj/wVAxGAFOTF396XKATGAp+HRCqJ81Ry60CNK1jDzk8dv6M6U\r\n"   \
-"HoS7RIFiM/9rXQCbJfiPD5xMTejZp5n5UYHAmxsxDaazfA5FuBhkfokKK6jD4Eq9\r\n"   \
-"1C94xGKb6X4/VkaPF7cqoBBw/bHxawXc0UEPjqayiBpCYU/rJoVZgLqFVP7Px3sv\r\n"   \
-"a1nOrNx8rPPI1hJ+ZOg8maiPTxHZnBVLakSSLQy/sWeWyazO1RnrbxjrbgQtYKz0\r\n"   \
-"e3nwGpu1w13vfckFmUSBhHXH7AAS/HpKC4IH7G2GAk3+n8iSSN71sZzpxonQwVbo\r\n"   \
-"pMZqLmbBm/7WPLc=\r\n"                                                   \
+#define TEST_SRV_CRT_RSA_SHA256                                         \
+"-----BEGIN CERTIFICATE-----\r\n"                                       \
```

Comment:
This is an ABI change: the variable `mbedtls_test_srv_crt_rsa` in libmbedx509 has changed size. Technically we should increase the soversion for that. Do we increase the soversion, decide that we don't care about ABI changes when they're only about test certificates which shouldn't have been in libmbedx509 in the first place, or keep the old certificates and correct the comment to match?

Reply:
This is a backport, I don't think we should increase the soversion.

Reply:
Oh good point. (Fortunately the CI caught that too.)

I agree that we don't want to increase the soversion in an LTS branch unless really absolutely necessary, which isn't the case here. (Note: in addition to violating stability guarantees, increasing the soversion in LTS branches would create problems as the next soversion is already used by current versions.)

I don't think we can just ignore the ABI change because it's only in something that shouldn't be part of the library in the first place (as much as I'd like to). The problem is user (especially linux distros) are running automated ABI checks, and ignoring that would raise a lot of red flags for them.

So I think the only realistic option is to keep the code as is, and change the comment to match the code instead.

Reply:
No, I was wrong here. This is not an ABI change. The ABI exposes a pointer to the array `mbedtls_test_srv_crt_rsa`, and it exposes a `size_t` constant containing the size of the array, but it does not expose the array itself. So it would have been fine to change the size.

(And then in https://github.com/ARMmbed/mbedtls/pull/3594 we update `tests/data_files/server2-sha256.crt` to have the same size in 2.7 as in later branches anyway.)

Reply:
Actually out ABI checking script would consider it to be a an ABI change. I think that's a limitation in [abi-dumper](https://github.com/lvc/abi-dumper) or [abi-compliance-checker](http://ispras.linux-foundation.org/index.php/ABI_compliance_checker), because I don't see how changing the size of the array could break an application.